pkg/provisioner/ironic/ironic.go

```diff
@@ -1760,34 +1761,60 @@ func (p *ironicProvisioner) RemoveBMCEventSubscriptionForNode(subscription metal
 	return operationComplete()
 }
 
-// TODO(hroyrh) : Replace with GetDataImageStatus function once the virtua_media.get
-// api is available.
-// Checks if the last VirtualMedia action(attach/detach) to a BareMetalHost was
-// successful of not.
-func (p *ironicProvisioner) IsDataImageReady() (isNodeBusy bool, nodeError error) {
-	// TODO(hroyrh)
-	// Get BareMetalHost VirtualMedia details using a GET api to vmedia
-
-	// Check if Ironic API version supports DataImage API
-	// Needs version >= 1.89
-	if !p.availableFeatures.HasDataImage() {
-		return true, fmt.Errorf("ironic version=%d doesn't support DataImage API, needs version>=1.89", p.availableFeatures.MaxVersion)
+// Uses the Ironic Virtual Media Get API which synchronously fetches the
+// virtual media details for the given node.
+// We return only the bool isImageAttached because the url in the response
+// body is always the same - based on the node uuid, so it is not useful
+// for any comparison purpose.
+func (p *ironicProvisioner) GetDataImageStatus() (isImageAttached bool, err error) {
+	// Check if Ironic API version supports Virtual Media Get API
+	// Needs version >= 1.93
+	if !p.availableFeatures.HasVirtualMediaGetAPI() {
+		return false, fmt.Errorf("ironic version=%d doesn't support Virtual Media Get API, needs version>=1.93", p.availableFeatures.MaxVersion)
 	}
 
 	node, err := p.getNode()
 	if err != nil {
 		return true, err
 	}
 
-	isNodeBusy = node.Reservation != ""
+	// If the node is busy, return error, since the get request will anyways
+	// fail
+	// TODO(hroyrh): is there a better way to do this ?
+	isNodeBusy := node.Reservation != ""
+	if isNodeBusy {
+		return false, fmt.Errorf("node is busy, retry GetVirtualMedia request in some time")
+	}
+
+	err = nodes.GetVirtualMedia(p.ctx, p.client, p.nodeID).Err
+	if err != nil {
+		return false, err
+	}
+
+	body := nodes.GetVirtualMedia(p.ctx, p.client, p.nodeID).Body
 
-	// In case the error node encountered was related to something else
-	// TODO(hroyrh) : Is this check valid ?
-	if !strings.Contains(node.LastError, "attach") && !strings.Contains(node.LastError, "detach") {
-		return isNodeBusy, nil
+	data, err := json.Marshal(body)
```

Comment:
All this JSON logic must be handled by GopherCloud, not us. You need to use Extract.

Reply:
Same as above, can we do it once I have added the extract function in gophercloud ?

Reply:
Are we ok with going with json.Marshal while we add the Extract to gophercloud?

Reply:
Let's at least leave a FIXME

Reply:
Also... why are we doing a Marschal followed by Unmarschall? What is Body even?

Reply:
result := nodes.GetVirtualMedia(p.ctx, p.client, p.nodeID)

returns the following:

{{[map[image: inserted:false media_types:[Floppy USBStick]] map[image:http://<address>/redfish/boot-12345678-aaaa-5856-bfb6-6aaaacd3dd11.iso inserted:true media_types:[CD DVD]]] 0 map[Content-Length:[212] Content-Type:[application/json] Date:[Fri, 28 Feb 2025 01:26:50 GMT] Openstack-Request-Id:[req-...] X-Openstack-Ironic-Api-Maximum-Version:[1.95] X-Openstack-Ironic-Api-Minimum-Version:[1.1] X-Openstack-Ironic-Api-Version:[1.93]] <nil>}}

So yeah we probably don't need the Unmarshall, seem unnecessary, we are already able to access Body and Err, so it goes down having the proper structs to handle the information.

Reply:
I've used VirtualMedia struct to capture the body data, will submit a patch to gophercloud to move it there.